repair/repair.cc

```diff
@@ -2106,8 +2106,21 @@ future<> repair_service::repair_tablets(repair_uniq_id rid, sstring keyspace_nam
             rlogger.debug("repair[{}] Table {}.{} does not exist anymore", rid.uuid(), keyspace_name, table_name);
             continue;
         }
-        // FIXME: we need to wait for current tablet movement and disable future tablet movement
-        auto erm = t->get_effective_replication_map();
+        locator::effective_replication_map_ptr erm;
+        while (true) {
+            _repair_module->check_in_shutdown();
+            erm = t->get_effective_replication_map();
+            const locator::tablet_map& tmap = erm->get_token_metadata_ptr()->tablets().get_tablet_map(tid);
+            if (!tmap.has_transitions()) {
+                break;
+            }
+            rlogger.info("repair[{}] Table {}.{} has tablet transitions, waiting for topology to quiesce", rid.uuid(), keyspace_name, table_name);
+            erm = nullptr;
+            co_await container().invoke_on(0, [] (repair_service& rs) {
+                return rs._tsm.local().await_quiesced();
```

Comment:
> We do not have a task covering the repair preparation part

Maybe we should have one if quiescing topology may block for a long time and the user wants to abort repair.

@tgrabiec could repair block behind decommission or rf change, that might take hours?  or would it wait just behind one batch, taking, say, minutes?

Reply:
> Also, I'm concerned that await_quiesced would continue poll without preventing new topology changes to start.

Yes, but to fix that, we'd have to implement the original design in which repairs requests are stored in group0 and tablet scheduler can see them as preemption request.

> @tgrabiec could repair block behind decommission or rf change, that might take hours? or would it wait just behind one batch, taking, say, minutes?

It will block behind whole decommission.

Reply:
> > Also, I'm concerned that await_quiesced would continue poll without preventing new topology changes to start.
> 
> Yes, but to fix that, we'd have to implement the original design in which repairs requests are stored in group0 and tablet scheduler can see them as preemption request.

Okay.  So it's work for a followup.

> > @tgrabiec could repair block behind decommission or rf change, that might take hours? or would it wait just behind one batch, taking, say, minutes?
> 
> It will block behind whole decommission.

I see.  As follow up, it would be good to make it by an async, abortable task.
Then, assuming repairs will be scheduled in single-tablet resolution, they could happen before migration for decommission, or after, or if we go very crazy, we can do repair-based migration instead of regular migration (but I don't expect that to be done any time soon, or ever).

Reply:
> Also, can we wait until the specific table is quiesced rather than waiting for the whole topology state machine to stop being busy?

We could, but it's an optimization I'm not sure is worth it. If there is an ongoing topology operation, it probably affects many tables, so it may not help much in practice.